src/builder-manifest.c

```diff
@@ -3802,12 +3786,57 @@ builder_manifest_install_dep (BuilderManifest *self,
 
   if (!builder_maybe_host_spawnv (NULL, NULL, 0, error, (const char * const *)args->pdata))
     {
-      g_autofree char *commandline = flatpak_quote_argv ((const char **)args->pdata);
-      g_prefix_error (error, "running `%s`: ", commandline);
       return FALSE;
     }
+  else
+    {
+      return TRUE;
+    }
+}
 
-  return TRUE;
+static gboolean
+builder_manifest_install_dep (BuilderManifest *self,
+                              BuilderContext  *context,
+                              char *const *remotes,
+                              gboolean opt_user,
+                              const char *opt_installation,
+                              const char *runtime,
+                              const char *version,
+                              gboolean opt_yes,
+                              GError         **error)
+{
+  g_autofree char *ref = NULL;
+  g_autofree char *commit = NULL;
+
+  if (version == NULL)
+    version = builder_manifest_get_runtime_version (self);
+
+  ref = flatpak_build_untyped_ref (runtime,
+                                   version,
+                                   builder_context_get_arch (context));
+
+  commit = flatpak_info (opt_user, opt_installation, "--show-commit", ref, NULL);
+
+  if (commit != NULL)
+    {
+      if (builder_manifest_install_single_dep(ref, NULL, opt_user, opt_installation,
+                                              error))
+      {
+        return TRUE;
+      }
+    }
+  else
+    {
+      for(const char *remote = *remotes; remote != NULL; remote = *(++remotes))
+	{
+	  if (builder_manifest_install_single_dep(ref, remote, opt_user, opt_installation,
+		         error))
```

Comment:
You can't pass error multiple times like this. If any of the previous calls in the loop has failed, then error is non-null, and then if *this* call fails you'll get a g_error warning about trying to set an already error:ed GError.

You need to use a new local_error here every time, and then save a copy of one of the errors to return from the loop if nothing succeeded. I suggest keeping the first error, that is traditional.

Reply:
Yes, glib ranted about this to me once already when I had accidental infinite loop. Will fix this.

Reply:
I *think* I now got it to work right. Tested by passing flathub as remote multiple times when refs were only available from freedesktop-sdk test repo.